app/controllers/sessions_controller.rb

```diff
@@ -0,0 +1,38 @@
+class SessionsController < ApplicationController
+  def show
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def edit
+  end
+
+  def create
+    @user = User.find_by_email(params[:email])
+
+    if @user && @user.authenticate(params[:password])
+      sign_in!(@user)
+      redirect_to root_url, notice: 'Logged in!'
+    else
+      ### IS THERE A BETTER WAY TO SHOW A LOGIN ERROR?
+      @user = User.new
+      @user.errors.add(:base, 'That email and password were not valid! Please try again.')
```

Comment:
@jfarmer Hey, just to confirm, do you mean that a controller should not be adding errors to a model like @user.errors.add? That instead the User model should have a method for adding an error? e.g. @user.add_error(:field, message)?